docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+
+```csharp
+class DemoComWrappers : ComWrappers
+{
+    protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count) =>
+        throw new NotImplementedException();
+
+    protected override object? CreateObject(IntPtr externalComObject, CreateObjectFlags flags) =>
+        throw new NotImplementedException();
+
+    protected override void ReleaseObjects(IEnumerable objects) =>
+        throw new NotImplementedException();
+}
+```
+
+To implement the `ComputeVtables()` method, decide which managed types you'd like to support. For this tutorial, we'll support the two previously defined interfaces (`IDemoGetType` and `IDemoStoreType`) and a managed type that implements the two interfaces (`DemoImpl`).
+
+```csharp
+class DemoImpl : IDemoGetType, IDemoStoreType
+{
+    string? _string;
+    public string? GetString() => _string;
+    public void StoreString(int _, string? str) => _string = str;
+}
+```
+
+For the `CreateObject()` method, you'll also need to determine what you'd like to support. In this case, though, we only know the COM interfaces we're interested in, not the COM classes. The interfaces being consumed from the COM side are the same as the ones we're projecting from the .NET side (that is, `IDemoGetType` and `IDemoStoreType`).
+
+We won't implement `ReleaseObjects()` in this tutorial.
+
+### Step 2 &ndash; Implement `ComputeVtables()`
+
+Let's start with the Managed Object Wrapper &ndash; these wrappers are easier. You'll build a [Virtual Method Table][wiki_vtable], or *vtable*, for each interface in order to project them into the COM environment. For this tutorial, you'll define a vtable as a sequence of pointers, where each pointer represents an implementation of a function on an interface &ndash; order is _very_ important here. In COM, every interface inherits from [`IUnknown`][api_iunknown]. The `IUnknown` type has three methods defined in the following order: `QueryInterface()`, `AddRef()`, and `Release()`. After the `IUnknown` methods come the specific interface methods. For example, consider `IDemoGetType` and `IDemoStoreType`. Conceptually, the vtables for the types would look like the following:
+
+```
+IDemoGetType    | IDemoStoreType
+==================================
+QueryInterface  | QueryInterface
+AddRef          | AddRef
+Release         | Release
+GetString       | StoreString
+```
+
+Looking at `DemoImpl`, we already have an implementation for `GetString()` and `StoreString()`, but what about the `IUnknown` functions? How to [implement an `IUnknown` instance][doc_impliunknown] is beyond the scope of this tutorial, but it can be done manually in `ComWrappers`. However, in this tutorial, you'll let the runtime handle that part. You can get the `IUnknown` implementation using the [`ComWrappers.GetIUnknownImpl()`][api_comwrappers] method.
```

Comment:
Is it intentional that the link for `GetIUnknownImpl` points to the more generic `ComWrappers` reference, why not the method page itself?

Reply:
Another piece of feedback is on the API reference for this method - it says very little. Would be nice to have some description of what the runtime implementation does (and roughly how) - specifically the QI behavior is interesting I think.

Reply:
> description of what the runtime implementation does

That is fair. I will add some notes.

> (and roughly how)

The how is an implementation detail I'd rather not document as it is always subject to change. Also, it shouldn't be relevant in the majority of cases.

Reply:
> The how is an implementation detail I'd rather not document as it is always subject to change. Also, it shouldn't be relevant in the majority of cases.

I completely agree that it's an implementation detail - I was just curious :-)

Reply:
It is just a linear search in two sets of arrays. The implementation is relatively simple and documented: https://github.com/dotnet/runtime/blob/404a89f843b121dfb3909c645eda80cc3a2aa9cc/src/coreclr/interop/comwrappers.cpp#L370-L456. So much easier than the built-in system...